File: ios/Pets/ViewModels/Sources/ViewModels/Favorites/FavoritesViewModel.swift
//
//  File.swift
//  
//
//  Created by Dylan Velez on 11/08/21.
//

import Foundation
import Architecture
import Domain
import Combine
import Models

public final class FavoritesViewModel: ViewModel {
    public typealias Event = FavoritesEvent
    public typealias FavoriteStore = Store<FavoritesState, FavoritesEvent,  FavoritesReducer>
    
    public struct ViewState {
        public var list: Pets = []
        public var isLoading = false
    }
    
    public var store: FavoriteStore
    private var stateCancellable: AnyCancellable?
    @Published private(set) public var state: ViewState = ViewState()
    
    public init(store: FavoriteStore) {
        self.store = store
        self.stateCancellable = store.$state.sink(receiveValue: stateChanged)
    }
    
    public func stateChanged(newState: FavoritesState) {
        self.state.list = newState.list.sorted { $0.id < $1.id }
        self.state.isLoading = newState.isLoading
    }
}

File: android/Pets/app/src/main/java/com/pets/viewModels/favorites/FavoritesViewModel.kt
package com.pets.viewModels.favorites

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.pets.architecture.stores.Store
import com.pets.architecture.viewModels.ViewModel
import com.pets.domain.favorites.FavoritesEvent
import com.pets.domain.favorites.FavoritesReducer
import com.pets.domain.favorites.FavoritesState
import com.pets.models.Pet

class FavoritesViewModel: ViewModel<FavoritesState, FavoritesEvent, FavoritesViewModel.ViewState>,
androidx.lifecycle.ViewModel() {
    data class ViewState(val list: List<Pet> = emptyList(),
                         val isLoading: Boolean = true)

    override var state: MutableLiveData<ViewState> = MutableLiveData()
    override val store: Store<FavoritesState, FavoritesEvent> =
        Store(MutableLiveData(FavoritesState()), reducer = FavoritesReducer())

    private val observer: Observer<FavoritesState> = Observer { onStateChange(it) }

    init {
        store.state.observeForever(observer)
    }

    override fun onCleared() {
        super.onCleared()
        store.state.removeObserver(observer)
    }

    override fun onStateChange(newState: FavoritesState) {
        state.value = ViewState(list = newState.list, isLoading = newState.isLoading)
    }
}